File: feature-map/src/main/java/com/levnovikov/feature_map/MapView.java
package com.levnovikov.feature_map;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;

import com.levnovikov.feature_map.lifecycle.MapLifecycleEvent;
import com.levnovikov.feature_map.lifecycle.MapLifecycleListener;

import javax.inject.Inject;

import io.reactivex.Observable;

/**
 * Author: lev.novikov
 * Date: 2/1/18.
 */

public class MapView extends com.google.android.gms.maps.MapView implements MapLifecycleListener {

    public MapView(@NonNull Context context) {
        super(context);
    }

    public MapView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public MapView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Inject
    MapInteractor interactor;

    @Inject
    Observable<MapLifecycleEvent> mapLifecycleStream;

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        interactor.onGetActive();
        initMap();
    }

    private void initMap() {
        getMapAsync(interactor);
        mapLifecycleStream //TODO unsubscribe
                .subscribe(event -> {
                    switch (event) {
                        case CREATE:
                            onCreate(event.getBundle());
                            break;
                        case RESUME:
                            onResume();
                            break;
                        case PAUSE:
                            onPause();
                            break;
                        case DESTROY:
                            onDestroy();
                            break;
                    }
                }, e -> { });
    }
}

File: feature-map/src/main/java/com/levnovikov/feature_map/MapView.kt
package com.levnovikov.feature_map

import android.content.Context
import android.util.AttributeSet
import com.levnovikov.feature_map.lifecycle.MapLifecycleEvent
import com.levnovikov.feature_map.lifecycle.MapLifecycleListener
import io.reactivex.Observable
import javax.inject.Inject

/**
 * Author: lev.novikov
 * Date: 2/1/18.
 */

class MapView : com.google.android.gms.maps.MapView, MapLifecycleListener {

    @Inject
    lateinit var interactor: MapInteractor

    @Inject
    lateinit var mapLifecycleStream: Observable<MapLifecycleEvent>

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {}

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        interactor.restoreState()
        initMap()
    }

    private fun initMap() {
        getMapAsync(interactor)
        mapLifecycleStream //TODO unsubscribe
        .subscribe({ event ->
            when (event) {
                MapLifecycleEvent.CREATE -> onCreate(event.bundle)
                MapLifecycleEvent.RESUME -> onResume()
                MapLifecycleEvent.PAUSE -> onPause()
                MapLifecycleEvent.DESTROY -> onDestroy()
            }
        }) { e -> }
    }
}
